### ISSUE TITLE
cannot find shellcheck for Raspian

### ISSUE THREAD

**Issue by NewtownGal** [koalaman/shellcheck#1270]
I see from the recent update to the online version of shellcheck that some of my scripts are too large for it to handle online. I tried to install it on Raspian, using apt-get update; apt-get install shellcheck; but it cannot find it. Do I need to install an additional path in the list of available sources ? Can I run it from the command line ?

Thank you in advance. It is a fantastic tool and has saved me lots of time. I'm having withdrawal symptoms now that some of my scripts have grown too large for the online version.

**Comment by koalaman** [comment]
I've been working on an automated ARM build to make this easier for you and others, but it's a bit tricky:

* Raspbian's GHC version is 7.6.3 from five years ago. The ecosystem has outgrown it, plus this version has a known optimizer issue  (#1131).
* ARMv5 (armel) support appears to have been broken since GHC 7.10.something. 

Which CPU or RPi version are you hoping to run this on? Would an ARMv7 (armhf) build work for you?

**Comment by NewtownGal** [comment]
Yes, I would run it on a Pi-2, which is an armv71.

**Comment by koalaman** [comment]
I tried this on a real Pi3, and it turns out that the latest version of Raspbian (Stretch) actually does have ShellCheck 0.4.4 packaged, and it also comes with an up-to-date GHC 8.0.2 that was able to compile ShellCheck from source (though it took three hours). The resulting binary was even armv6 compatible according to qemu.

Which Raspbian version did you try this on?

I did manage to get an armv7hf cross-compiler running in Docker, but I'm still annoyed that I wasn't able to get it to build a version compatible with Pi1/PiZero, especially now that Raspbian does it so effortlessly. If I can't get my crosscompiler to produce armv6, maybe I'll just emulate their entire toolchain.

**Comment by enjunear** [comment]
Hi guys, just noticed this issue after managing to get shellcheck working on a raspberry pi 3.

In my case, I'm using shellcheck in GitLab CI. When running on a GitLab hosted runner, using the `koalaman/shellcheck-alpine` works great, but not when running on my raspberry pi 3 based runner.

For my raspberry pi 3, I created a new `.gitlab-ci-arm.yml` file, using the arm32v7 debian image as follows:

```yml
shellcheck:
  image: arm32v7/debian
  script:
    - apt-get -q update && apt-get -q install shellcheck -y
    - shellcheck *.sh
```

It may not be the most minimal solution, but it works. I'm not sure if it helps your use case @NewtownGal

**Comment by koalaman** [comment]
In addition to the ShellCheck packaged for Raspbian Stretch, there is now also an [official armv6hf build](https://storage.googleapis.com/shellcheck/shellcheck-stable.linux.armv6hf.tar.xz).

This build is available with the `latest`, `stable` and `v0.5.0` tags just like the `x86_64` builds.

For anyone interested, this is based on [resin.io Raspbian](https://hub.docker.com/r/resin/rpi-raspbian/) containers using qemu user emulation to run a armv6hf userland that compiles ShellCheck using an armv6hf GHC.

I obviously hoped to do it through cross-compilation, but this was a massive pain that required a pile of hacks for armv7 on the latest git branch of ghc, and even after spending my week on that, I never managed to get it to successfully build an armv6 compatible binary that would work on Raspberry Pi One or Zero.

I've only confirmed this on a Pi3 that i had lying around, plus qemu in `arm1176` mode. Testing on Pi2, Pi1 and Pi0 would be much appreciated.
